### PR TITLE
[WIP] Syft <> Lightning

### PR DESCRIPTION
**Pull request by tudorcebere**
## What does this PR do?

Work in progress integration of Lightning with Syft

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM ! We will have to introduce a Secure Plugin to remove those checks.

**Review by justusschock — CHANGES_REQUESTED** [review]
LGTM in general. We should really try to encapsulate this in a plugin later on.

Can you add tests for that since basically every line guarded by `_PYSYFT_AVAILABLE` is not tested.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -302,7 +303,11 @@ def _save_model(self, filepath: str, trainer, pl_module):
 
         # delegate the saving to the trainer
         if self.save_function is not None:
-            self.save_function(filepath, self.save_weights_only)
+            # Todo (tudorcebere): Add support for saving the model   # noqa E265
+            if _PYSYFT_AVAILABLE:
+                pass
+            else:
+                self.save_function(filepath, self.save_weights_only)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:310]: This means no models will be saved if syft is installed at all, regardless of it it's being used

Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:310]: Yes, it isn't supported yet.

Comment (justusschock) [reply pytorch_lightning/callbacks/model_checkpoint.py:309]: But I agree with @carmocca that this shouldn't be the right type of check here. We should somehow check if it's used not if it's just installed

Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:309]: You are entirely right ! Let me work on it.

File: pytorch_lightning/plugins/secure/pysyft.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+import torch
+from torch.nn import Module
+from torch.optim import Optimizer
+
+from pytorch_lightning.plugins.secure.base import BaseSecurePlugin
+from pytorch_lightning.utilities.imports import _PYSYFT_AVAILABLE
+
+
+class PySyftPlugin(BaseSecurePlugin):
```
Comment (justusschock) [inline pytorch_lightning/plugins/secure/pysyft.py:24]: is all this stuff used already?

**Review by justusschock — APPROVED** [review]
LGTM. Can we still move the support here to experimental, since this is likely to change i. future iterations?

File: pytorch_lightning/plugins/secure/pysyft.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/plugins/secure/pysyft.py:24]: Done !

File: pytorch_lightning/experimental/plugins/secure/pysyft.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        sy_model.device = model_ref.device
+        sy_model.hparams = model_ref.hparams
+        sy_model.on_save_checkpoint = model_ref.on_save_checkpoint
+        trainer.training_type_plugin.model = sy_model
+        trainer.accelerator.optimizer_state = PySyftPlugin.optimizer_state
+        trainer.save_checkpoint(filepath, save_weights_only)
+        trainer.training_type_plugin.model = model_ref
+
+
+if _PYSYFT_AVAILABLE:
```
Comment (Borda) [inline pytorch_lightning/experimental/plugins/secure/pysyft.py:47]: this all does not need to be in if, just the needed sections

File: pytorch_lightning/setup_tools.py [hunk 4]
```diff
@@ -35,6 +35,8 @@ def _load_requirements(path_dir: str, file_name: str = 'requirements.txt', comme
         # skip directly installed dependencies
         if ln.startswith('http'):
             continue
+        if ln.startswith('git'):
```
Comment (Borda) [inline pytorch_lightning/setup_tools.py:38]: I would prefer to install with zip as git version clones all repo with history

**Review by Borda — COMMENTED** [review]
I would not create an experimental package as will yield in a move to stable repo later and breaking imports for the user...

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 5]
```diff
@@ -301,7 +302,11 @@ def _save_model(self, filepath: str, trainer):
 
         # delegate the saving to the trainer
         if self.save_function is not None:
-            self.save_function(filepath, self.save_weights_only)
+            if _PYSYFT_AVAILABLE:
+                from pytorch_lightning.experimental.plugins.secure.pysyft import PySyftPlugin
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:306]: > I would not create an experimental package as will yield in a move to stable repo later and breaking imports for the user...

That's the point of `experimental` folder. We develop it there, and when ready we had BC section to the changelog as PyTorch does.

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:306]: Screenshot 2021-03-05 at 12 53 52

**Comment by tchaton** [comment]
> I would not create an experimental package as will yield in a move to stable repo later and breaking imports for the user...

That's the point of experimental folder. We develop it there, and when ready we moved it elsewhere. The experimental is used to clearly share with the users this feature will change or move.

When ready, we announce in the release there is Backward Imcompatible changes as PyTorch does.

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:306]: yes.

File: pytorch_lightning/experimental/plugins/secure/pysyft.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/experimental/plugins/secure/pysyft.py:47]: Done !

File: pytorch_lightning/experimental/plugins/secure/pysyft.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        def setup(self, stage_name: str):
+            self.get = self.module.get
+            self.send = self.module.send
+            self.send_model()
+
+        def is_remote(self) -> bool:
+            # Training / Evaluation is done remotely and Testing is done locally unless run_locally is True
+            if self.run_locally or (not self.trainer.training and self.trainer.evaluation_loop.testing):
+                return False
+            return True
```
Comment (carmocca) [inline pytorch_lightning/experimental/plugins/secure/pysyft.py:83]:
```suggestion
            return not self.run_locally and not self.trainer.testing
```

File: pytorch_lightning/experimental/plugins/secure/pysyft.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+        @property
+        def model(self) -> SyModelProxyType:
+            if self.is_remote():
+                return self.remote_model
+            else:
+                if self.download_back:
+                    return self.get_model()
+                else:
+                    return self.module
```
Comment (carmocca) [inline pytorch_lightning/experimental/plugins/secure/pysyft.py:97]:
```suggestion
            if self.download_back:
                return self.get_model()
            return self.module
```

File: pytorch_lightning/experimental/plugins/secure/pysyft.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+        def send_model(self) -> None:
+            self.remote_model = self.module.send(self.duet)
+
+        def get_model(self) -> type(Module):  # type: ignore
```
Comment (carmocca) [inline pytorch_lightning/experimental/plugins/secure/pysyft.py:102]: why is the return type not just `Module`?

File: pytorch_lightning/experimental/plugins/secure/pysyft.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    SyModuleProxyType = Union[ModuleType, Module]
+    SyModelProxyType = Union[Module, sy.Module]
+
+    # cant use lib_ast during test search time
+    TorchTensorPointerType = Any  # sy.lib_ast.torch.Tensor.pointer_type
+    SyTensorProxyType = Union[torch.Tensor, TorchTensorPointerType]  # type: ignore
+
+    class SyLightningModule(LightningModule):
+
+        def __init__(self, download_back: bool = False, run_locally: bool = False) -> None:
```
Comment (carmocca) [inline pytorch_lightning/experimental/plugins/secure/pysyft.py:65]: where is `self.module` defined? shouldn't `module` be an init argument?

File: pytorch_lightning/experimental/plugins/secure/pysyft.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            super().__init__()
+            # Those are helpers to easily work with `sy.Module`
+            self.duet = sy.client_cache["duet"]
+            self.remote_torch = sy.client_cache["duet"].torch
+            self.local_torch = globals()["torch"]
+            self.download_back = download_back
```
Comment (carmocca) [inline pytorch_lightning/experimental/plugins/secure/pysyft.py:71]: what is download_back for?

File: pytorch_lightning/trainer/supporters.py [hunk 11]
```diff
@@ -101,7 +113,10 @@ def _agg_memory(self, how: str):
             if self.rotated:
                 return getattr(self.memory, how)()
             else:
-                return getattr(self.memory[:self.current_idx], how)()
+                if _PYSYFT_AVAILABLE:
+                    return getattr(self.memory[list(range(0, self.current_idx))], how)()
+                else:
+                    return getattr(self.memory[:self.current_idx], how)()
```
Comment (carmocca) [inline pytorch_lightning/trainer/supporters.py:119]:
```suggestion
            if _PYSYFT_AVAILABLE:
                return getattr(self.memory[list(range(0, self.current_idx))], how)()
            return getattr(self.memory[:self.current_idx], how)()
```

File: pytorch_lightning/trainer/training_loop.py [hunk 12]
```diff
@@ -344,8 +353,15 @@ def _process_training_step_output(self, training_step_output, split_batch):
         # -----------------------------------------
         # no need for these checks in 1.0.0
         # TODO: remove checks in 1.0.0
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:355]: Unrelated to this PR but we should address this...

File: pytorch_lightning/utilities/apply_func.py [hunk 13]
```diff
@@ -155,6 +155,14 @@ def batch_to(data):
             return device_data
 
         kwargs = dict(non_blocking=True) if isinstance(data, torch.Tensor) else {}
+
+        if _PYSYFT_AVAILABLE:
+            from syft.core.pointer.pointer import Pointer
+            from syft import client_cache
+            if isinstance(data, Pointer):
+                data = data.get(request_block=True, delete_obj=False)
+                data = data.send(client_cache["duet"])
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:164]: why do we need to send here?

File: pytorch_lightning/utilities/imports.py [hunk 14]
```diff
@@ -71,3 +71,4 @@ def _compare_version(package: str, op, version) -> bool:
 _TORCHTEXT_AVAILABLE = _module_available("torchtext")
 _TORCHVISION_AVAILABLE = _module_available('torchvision')
 _XLA_AVAILABLE = _module_available("torch_xla")
+_PYSYFT_AVAILABLE = _module_available("syft")
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:74]: nit: these variables are sorted

File: requirements/extra.txt [hunk 15]
```diff
@@ -7,4 +7,5 @@ torchtext>=0.5
 onnx>=1.7.0
 onnxruntime>=1.3.0
 hydra-core>=1.0
+git+https://github.com/OpenMined/PySyft.git@refs/pull/5221/head
```
Comment (carmocca) [inline requirements/extra.txt:10]: is there a plan to merge that PR before this one?

File: pytorch_lightning/experimental/plugins/secure/pysyft.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/experimental/plugins/secure/pysyft.py:79]: Done !

File: pytorch_lightning/experimental/plugins/secure/pysyft.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/experimental/plugins/secure/pysyft.py:93]: Done !

File: pytorch_lightning/experimental/plugins/secure/pysyft.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/experimental/plugins/secure/pysyft.py:98]: No idea. Not my code ahah. Remove it.

File: pytorch_lightning/trainer/supporters.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py]: Done !

File: pytorch_lightning/experimental/plugins/secure/pysyft.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/experimental/plugins/secure/pysyft.py:67]: Renamed `request_parameters`. It also to pull the trained weights from the remote machine.

File: pytorch_lightning/trainer/training_loop.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:355]: Yes.

File: pytorch_lightning/utilities/apply_func.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/utilities/apply_func.py:171]: No idea. That's from OpenMined. @tudorcebere

File: pytorch_lightning/utilities/imports.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/utilities/imports.py:74]: Done !

File: requirements/extra.txt (hunk 15 above)
Comment (tchaton) [reply requirements/extra.txt:10]: Possibly.

File: pytorch_lightning/utilities/imports.py [hunk 16]
```diff
@@ -71,3 +71,13 @@ def _compare_version(package: str, op, version) -> bool:
...
+
+
+def is_syft_initialized():
+    if _PYSYFT_AVAILABLE:
+        import syft as sy
+        from syft.core.node.vm.client import VirtualMachineClient
+        if "duet" in sy.client_cache and isinstance(sy.client_cache["duet"], VirtualMachineClient):
+            return True
+    return False
```
Comment (carmocca) [inline pytorch_lightning/utilities/imports.py:83]: Move this into pytorch_lightning/experimental/plugins/secure/pysyft.py?

File: requirements/extra.txt (hunk 15 above)
Comment (carmocca) [reply requirements/extra.txt:10]: i wouldn't merge this until the PR is merged.

Or at least, i'd ask their team to create a branch in their repository with these changes. This would also allow installing a `.zip` as @Borda requested

**Review by ananthsub — COMMENTED** [review]
@tudorcebere @tchaton is there a high-level description you can add for what's required for pysyft integration? what's needed for followup PRs? gating by `is_syft_initialized()` in the main module/trainer is a short-term solution.

does/should the base secure plugin interact with the training type plugin?

**Comment by tchaton** [comment]
> @tudorcebere @tchaton is there a high-level description you can add for what's required for pysyft integration? what's needed for followup PRs? gating by `is_syft_initialized()` in the main module/trainer is a short-term solution.
> 
> does/should the base secure plugin interact with the training type plugin?

Hey @ananthsub. Excellent questions ! We don't have a high-level description yet. I plan to work on it next week. To make it work, you just need to install the PySyft for the branch specified and run the syft_test.

As you have seen there is still some nits to clean out in up-coming PRs. Right now, `is_syft_initialized ` is needed for know if the duet object has been created and might be moved to PySyft directly when they clean this logic internally.

We don't know yet what a secure plugin should do, but we think it should live in the training type plugin or be one by itself.
Working with Opacus | Flower should help us to find the API for Federated Learning use-cases.. 

Best,
T.C

File: pytorch_lightning/utilities/apply_func.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/utilities/apply_func.py:171]: It is used to cash data from my understanding.

Comment (tudorcebere) [reply pytorch_lightning/utilities/apply_func.py:171]: It works as a cast, we will come with this mechanism as soon as possible, but we want to get the real pointer type here.
